Include/internal/pycore_lock.h

```diff
@@ -213,6 +213,41 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
     return _PyOnceFlag_CallOnceSlow(flag, fn, arg);
 }
 
+// A readers-writer (RW) lock. The lock supports multiple concurrent readers or
+// a single writer. The lock is write-preferring: if a writer is waiting, then
+// new readers will be blocked. This avoids starvation of writers.
+//
+// The bitfield is structured as follows:
+//
+// N ...                                     2  1  0
+// +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
+// |                  reader_count           |P |WL|
+// +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
+//
+// The least significant bit (WL / _Py_WRITE_LOCKED) indicates if the mutex is
+// write-locked. The next bit (PK/ _Py_HAS_PARKED) indicates if there are
+// parked threads (either readers or writers). The remaining bits
+// (reader_count) indicate the number of readers holding the lock.
+//
+// Note that reader_count must be zero if WL is set, and vice versa. The lock
+// can only be held by readers or a writer, but not both.
+//
+// The design is optimized for simplicity of the implementation. The lock is
+// not fair: if fairness is desired, use an additional PyMutex to serialize
+// writers. The lock is also not reentrant.
+typedef struct {
+    uintptr_t bits;
+} _PyRWMutex;
```

Comment:
I would prefer to avoid the type prefixes because they add a lot of verbosity, which makes it harder to see the important parts of the algorithm. They'll force a lot of extra line wrapping for otherwise simple operations.

I added `_Py_WRITE_LOCKED` because  "locked" alone is ambiguous in this context -- that bit specifically means write-locked (as opposed to read-locked).

When MSVC finally makes C11 atomics non-experimental, we can probably use a bitfield for `_PyRWMutex` (and most other structs here), which I think will make things both more clear and concise.

Reply:
I still think the prefixed names would be clearer, but I get what you're saying and I'm fine with the way things are.